server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankShardContext.java

```diff
@@ -0,0 +1,158 @@
+            // otherwise mark top docs as empty
+            if (searchContext.suggest() != null
+                || searchContext.trackTotalHitsUpTo() != TRACK_TOTAL_HITS_DISABLED
+                || searchContext.aggregations() != null) {
+                QueryPhase.executeInternal(rrfRankSearchContext);
+            } else {
+                searchContext.queryResult()
+                    .topDocs(
+                        new TopDocsAndMaxScore(
+                            new TopDocs(new TotalHits(0, TotalHits.Relation.EQUAL_TO), Lucene.EMPTY_SCORE_DOCS),
+                            Float.NaN
+                        ),
+                        new DocValueFormat[0]
+                    );
+            }
+
+            List<TopDocs> rrfRankResults = new ArrayList<>();
+            rrfRankSearchContext.windowSize(windowSize);
+            boolean searchTimedOut = querySearchResult.searchTimedOut();
+            long serviceTimeEWMA = querySearchResult.serviceTimeEWMA();
+            int nodeQueueSize = querySearchResult.nodeQueueSize();
+
+            // run each of the rrf queries
+            for (Query query : queries) {
+                // if a search timeout occurs, exit with partial results
+                if (searchTimedOut) {
+                    break;
+                }
+                rrfRankSearchContext.rrfRankQuery(query);
+                QueryPhase.executeInternal(rrfRankSearchContext);
+                QuerySearchResult rrfQuerySearchResult = rrfRankSearchContext.queryResult();
+                rrfRankResults.add(rrfQuerySearchResult.topDocs().topDocs);
+                serviceTimeEWMA += rrfQuerySearchResult.serviceTimeEWMA();
+                nodeQueueSize = Math.max(nodeQueueSize, rrfQuerySearchResult.nodeQueueSize());
+                searchTimedOut = rrfQuerySearchResult.searchTimedOut();
+            }
+
+            sort(rrfRankResults, querySearchResult);
+
+            // record values relevant to all queries
+            querySearchResult.searchTimedOut(searchTimedOut);
+            querySearchResult.serviceTimeEWMA(serviceTimeEWMA);
+            querySearchResult.nodeQueueSize(nodeQueueSize);
+        } catch (QueryPhaseExecutionException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new QueryPhaseExecutionException(searchContext.shardTarget(), "Failed to execute main query", e);
+        }
+    }
+
+    protected void sort(List<TopDocs> rrfRankResults, QuerySearchResult querySearchResult) {
+        // combine the disjointed sets of TopDocs into a single set or RRFRankDocs
+        // each RRFRankDoc will have both the position and score for each query where
+        // it was within the result set for that query
+        // if a doc isn't part of a result set its position will be NO_RANK [0] and
+        // its score is [0f]
+        int queries = rrfRankResults.size();
+        Map<Integer, RRFRankDoc> docsToRankResults = new HashMap<>();
```

Comment:
I think we could set expected size for this hashmap.

Reply:
Would you just have it be queries*window_size?

Reply:
`queries*window_size` I think so. `Maps.hashMapWithExpectedSize` or whatever its called